### PR TITLE
Support re-aliasing of type alias instantiations

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR continues the work in #42149 by supporting re-aliasing of type alias instantiations. For example:

```ts
type Thing<T> = { value: T };
type ArrayThing<U> = Thing<U[]>;
type StringArrayThing = ArrayThing<string>;

declare let t1: Thing<string[]>;     // Thing<string[]>
declare let t2: ArrayThing<string>;  // ArrayThing<string> (previously Thing<string[]>)
declare let t3: StringArrayThing;    // StringArrayThing (previously Thing<string[]>)

interface Item<T> { value: T }
type ArrayItem<U> = Item<U[]>;
type NumberArrayItem = ArrayItem<number>;

declare let t4: Item<number[]>;     // Item<number[]>
declare let t5: ArrayItem<number>;  // ArrayItem<number>
declare let t6: NumberArrayItem;    // NumberArrayItem (previously ArrayItem<number>)

function foo<T, K extends keyof T>(obj: T, key: K) {
    let x: Omit<T, K>;  // Omit<T, K> (previously Pick<T, Exclude<keyof T, K>>)
}
```

Previously, instantiations of aliased anonymous object, mapped, or conditional types with a new type alias would continue to display the original type alias in diagnostics, quick info, and declaration files. Now, re-aliasing of such types consistently adopts and displays the new type alias.

In addition, instantiations of distributive types (such as homomorphic mapped types and distributive conditional types) now properly preserve the original type alias when the instantiation results in a union type:

```ts
type A = { a: string };
type B = { b: string };

declare let p: Partial<A | B>;  // Partial<A | B> (previously Partial<A> | Partial<B>)
```

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster

**Comment by ahejlsberg** [comment]
@robpalme This is the continuation PR I mentioned in #42149. Give it a try and let me know how it goes.

**Comment by ahejlsberg** [comment]
Tests all look good, only changes are improved diagnostics. Performance is unaffected.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

File: tests/baselines/reference/conditionalTypes1.errors.txt [hunk 1]
```diff
@@ -185,22 +195,32 @@ tests/cases/conformance/types/conditional/conditionalTypes1.ts(288,43): error TS
...
         ~
-!!! error TS2322: Type 'Pick<T, NonFunctionPropertyNames<T>>' is not assignable to type 'Pick<T, FunctionPropertyNames<T>>'.
+!!! error TS2322: Type 'NonFunctionProperties<T>' is not assignable to type 'FunctionProperties<T>'.
 !!! error TS2322:   Type 'FunctionPropertyNames<T>' is not assignable to type 'NonFunctionPropertyNames<T>'.
+!!! error TS2322:     Type 'keyof T' is not assignable to type 'T[keyof T] extends Function ? never : keyof T'.
+!!! error TS2322:       Type 'string | number | symbol' is not assignable to type 'T[keyof T] extends Function ? never : keyof T'.
+!!! error TS2322:         Type 'string' is not assignable to type 'T[keyof T] extends Function ? never : keyof T'.
+!!! error TS2322:           Type 'keyof T' is not assignable to type 'never'.
+!!! error TS2322:             Type 'string | number | symbol' is not assignable to type 'never'.
+!!! error TS2322:               Type 'string' is not assignable to type 'never'.
```
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply tests/baselines/reference/conditionalTypes1.errors.txt:214]: Sounds like we can't avoid re-elaboration based on type IDs

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by robpalme** [comment]
@ahejlsberg This is even better than the previous attempt 🎉 

As before, there are no inlining/size regressions - I reviewed every changed line and this PR is a pure win.  This PR improves upon https://github.com/microsoft/TypeScript/pull/42149/commits/2c2d06dfe1d40bafc798a2f7c62e5662068738c8 by also handling aliases to `Record` and `Pick<obj, union>`.

```ts
// Previous: 2c2d06
  myRecordProp: Record<string, any>;
  myObjectProp: Pick<{
      lotsOfProps,
    }, "red" | "green" | "blue
  >;
```

...is replaced by...

```ts
// Master: Using PR#42284
  myRecordProp: MyRecordType;
  myObjectProp: MyColorsOnly<{
      lotsOfProps,
    }, Color
  >;
```

On the large package I referred to in [previous](https://github.com/microsoft/TypeScript/pull/42149#issuecomment-755292967) [reports](https://github.com/microsoft/TypeScript/pull/42149#issuecomment-757134847), this PR results in a further 10KB decrease vs the previous attempt, i.e the `Record` & `Pick` changes explains the 10KB win.  Meaning the net effect of this PR is now:

- 50 declaration files out of 394 were affected by this PR
- Net size of all declaration files shrank by 41KB
  - 50 files shrank (total size of these files reduced by 41KB from 585KB to 544KB)
  - 0 files grew

**Pull request merged**